src/pushprocessor.ts

```diff
@@ -123,6 +123,12 @@ export class PushProcessor {
      */
     public constructor(private readonly client: MatrixClient) {}
 
+    /**
+     * Maps the original key from the push rules to a list of property names
+     * after unescaping.
+     */
+    private readonly parsedKeys = new Map<string, string[]>();
```

Comment:
Originally I was going to add some sort of cached property to the `IEventMatchCondition`, but that's just an interface so this doesn't seem doable? Let me know if I'm wrong!

Downside of this approach is that if you delete a push rule then the key would still be cached.

Reply:
You could add a field to IEventMatchCondition which is optional, e.g. `memoizedCheckFn?(ev: MatrixEvent) => boolean` and set it on all such conditions as part of parsing/processing them whenever push rules change. Or have another interface which extends IEventMatchCondition e.g. `interface EventMatchConditionWithMemoizedHandler extends IEventMatchCondition` and "upgrade" them as part of processing push rules

Reply:
I figured there was a way there, but does it make sense to add it there? It seemed odd to add functionality to an interface like that.

I did consider the upgrade path, but it seemed potentially fragile, I'll take another look.

Reply:
Better than a leaky map :)

Could also prune the map when receiving new push rules via /sync

Reply:
If I understand what you mean, you'll be adding the data to the object that implements the interface, and adding code to the interface to manage that data, which both seem fine. If you'd like a call to talk through how to implement this more, ping me.

Reply:
> You could add a field to IEventMatchCondition which is optional

My fear with this is that if we somehow got a push rule with this field set we wouldn't calculate it. This might require a malicious homeserver, but 🤷 

> Or have another interface which extends IEventMatchCondition [...] and "upgrade" them as part of processing push rules whenever they change

This looks doable in `rewriteDefaultRules`, I think there might be some issues with multiple interfaces sharing the same `ConditionKind`? I did try creating a non-interface type which contained an `IEventMatchCondition` as a field, but adding that to `PushRuleCondition` caused issues with various methods which took an `IPushRuleCondition`.

I think we can do it in `rewriteDefaultRules` though and set a field, similar to the first approach. That seems simpler.